src/components/number-input/number-input.tsx

```diff
@@ -0,0 +1,253 @@
+
+        if(value !== next) {
+            this.committed = false;
+            this.setState({value: next});
+        }
+    }
+
+    private stepValue(e: React.SyntheticEvent<HTMLElement>, direction: Direction) {
+        const {value} = this.state;
+        const {step} = this.props;
+        const next = (direction == Direction.Increase ?
+            isNumber(value) ? value + step! : step! :
+            isNumber(value) ? value - step! : -step!);
+
+        this.commit(next);
+    }
+
+    private handleFocus: React.FocusEventHandler<HTMLElement> =
+        e => this.setState({focus: true});
+
+    private handleBlur: React.FocusEventHandler<HTMLElement> =
+        e => this.setState({focus: false});
+
+    private handleIncrement: React.MouseEventHandler<HTMLElement> =
+        e => this.stepValue(e, Direction.Increase);
+
+    private handleDecrement: React.MouseEventHandler<HTMLElement> =
+        e => this.stepValue(e, Direction.Decrease);
+
+    private handleInputKeyDown: React.KeyboardEventHandler<HTMLElement> =
+        e => {
+            switch (e.keyCode) {
+                case KeyCodes.UP:
+                    this.stepValue(e, Direction.Increase);
+                    e.preventDefault();
+                    break;
+                case KeyCodes.DOWN:
+                    this.stepValue(e, Direction.Decrease);
+                    e.preventDefault();
+                    break;
+                case KeyCodes.ENTER:
+                    this.commit(this.state.value);
+                    e.preventDefault();
+                    break;
+                case KeyCodes.ESCAPE:
+                    this.revert();
+                    e.preventDefault();
+                    break;
+            }
+        }
+
+    private handleInputBlur: React.FocusEventHandler<HTMLInputElement> =
+        e => this.commit(this.state.value);
+
+    private handleInputChange: React.ChangeEventHandler<HTMLInputElement> =
+        e => {
+            const {onInput} = this.props;
+            const value = e.target.value;
+            const next = value !== '' ?
+                Number(e.target.value) :
+                undefined;
+
+            this.updateValue(next);
+            onInput!(value);
+        }
+
+    componentWillReceiveProps({value}: NumberInputProps) {
+        const {min, max} = this.props;
+
+        if (value !== this.state.value) {
+            this.committed = true;
+            this.setState({value});
+            if (isNumber(value) && (value as number < min! || value > max!)) {
+                console.error('value is out of range');
+            }
+        }
+    }
+
+    render() {
+        const {value, focus, error} = this.state;
+        const {step, min, max, onInput, children, ...inputProps} = this.props;
+        const disableIncrement = inputProps.disabled || (isNumber(value) && value >= max!);
+        const disableDecrement = inputProps.disabled || (isNumber(value) && value <= min!);
```

Comment:
`isNumber(value)` asserts the type of `value` variable for the rest of the expression. Other configuration of this code look ugly. `disableIncrement` and `disableDecrement` have different semantics and it just so happens that they are differently determined by `<=` and `>=`. Extracting this into a function would artificially couple them.